src/main/java/bio/terra/service/duos/DuosService.java

```diff
@@ -0,0 +1,76 @@
+package bio.terra.service.duos;
+
+import bio.terra.model.DuosFirecloudGroupModel;
+import bio.terra.service.auth.iam.IamService;
+import bio.terra.service.auth.iam.exception.IamConflictException;
+import com.google.common.annotations.VisibleForTesting;
+import java.util.Optional;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class DuosService {
+  private static final Logger logger = LoggerFactory.getLogger(DuosService.class);
+
+  private final DuosDao duosDao;
+  private final IamService iamService;
+
+  @Autowired
+  public DuosService(DuosDao duosDao, IamService iamService) {
+    this.duosDao = duosDao;
+    this.iamService = iamService;
+  }
+
+  public Optional<DuosFirecloudGroupModel> retrieveFirecloudGroup(String duosId) {
+    return Optional.ofNullable(duosDao.retrieveFirecloudGroup(duosId));
+  }
+
+  public DuosFirecloudGroupModel createFirecloudGroup(String duosId) {
+    logger.info("Creating Firecloud group for {} users", duosId);
+
+    // First try with the more readable group name.
+    String groupName = constructFirecloudGroupName(duosId);
+    try {
+      iamService.createGroup(groupName);
+    } catch (IamConflictException ex) {
+      logger.warn(
+          "Firecloud group {} already exists: trying creation with a unique name", groupName);
+      groupName = constructUniqueFirecloudGroupName(duosId);
+      iamService.createGroup(groupName);
+    }
+    String groupEmail = iamService.getGroupEmail(groupName);
+    logger.info("Successfully created Firecloud group {} for {} users", groupName, duosId);
+
+    boolean inserted = duosDao.insertFirecloudGroup(duosId, groupName, groupEmail);
+    if (!inserted) {
+      iamService.deleteGroup(groupName);
+      throw new RuntimeException(
```

Comment:
I think we want a more specific exception (we generally only use RuntimeException in tests)